tt_metal/impl/device/device_pool.cpp

```diff
@@ -371,13 +374,97 @@ bool DevicePool::close_device(chip_id_t device_id) {
+                        // found the target device to shutdown, add all devices in this tunnel
+                        for (uint32_t i = t.size() - 1; i > 0; i--) {
+                            devices_to_close.push_back(t[i]);
+                        }
+                        break;
+                    }
+                }
+            }
+        }
+    }
+    // Loop over all devices and add any mmio devices
+    for (const auto dev : devices) {
+        if(dev->is_mmio_capable()) {
+            devices_to_close.push_back(dev->id());
+        }
+    }
+    // FIXME: ideally we can close mmio devices, and other active devices in the destructor of DevicePool.
+    // Was encountering issues with the dispatch_constants being destroyed before the DevicePool destructor,
+    // which leads to device->close() hitting asserts. We need to move the ownership of dispatch_constants
+    // to the device, so it doesn't go out of scope before the device is closed.
+    bool auto_close_mmio = is_galaxy;
+    // For Galaxy if an mmio device's tunnels are being closed, close the mmio device as well
+    if (is_galaxy) {
+        for (const auto dev : devices) {
+            bool auto_close_mmio = true;
+            const auto &mmio_device_id = tt::Cluster::instance().get_associated_mmio_device(dev->id());
+            auto mmio_dev_handle = tt::DevicePool::instance().get_active_device(mmio_device_id);
+            auto tunnels_from_mmio = mmio_dev_handle->tunnels_from_mmio_;
+            //iterate over all tunnels origination from this mmio device
+            for (auto t : tunnels_from_mmio) {
+                //iterate over all tunneled devices (tunnel stops) in this tunnel
+                for (uint32_t ts = t.size() - 1; ts > 0; ts--) {
+                    // If the device is active and not in the list of devices to close, don't auto close mmio
+                    if (this->is_device_active(t[ts]) and std::find(devices_to_close.begin(), devices_to_close.end(), t[ts]) == devices_to_close.end()) {
+                        auto_close_mmio = false;
+                        break;
+                    }
+                }
+            }
+            if (auto_close_mmio and std::find(devices_to_close.begin(), devices_to_close.end(), mmio_device_id) == devices_to_close.end()) {
+                devices_to_close.push_back(mmio_device_id);
+            }
+        }
+    }
+
+    // Global Sync across all devices that are being closed
+    // We need to ensure that commands sent to each device have been completed
+    // before closing any device + modifying routing info.
+    // If this is not done, non-blocking CCLs followed by a close will hang, since
+    // the main thread will modify device state while the CCL is running on device.
+    for (const auto dev_id : devices_to_close) {
+        auto *dev = tt::DevicePool::instance().get_active_device(dev_id);
+        dev->synchronize(); // Synchronize worker queue
```

Comment:
why do we need to call `dev->synchronize()` here but not in the `DevicePool::close_device` method?